nucypher/network/sensors.py

```diff
@@ -0,0 +1,186 @@
+    SAMPLE_SIZE = 1            # Ursulas
+    SENSOR_SENSITIVITY = 0.5   # Threshold
+    CHARGE_RATE = 0.9          # Measurement Multiplier
+
+    class Unreachable(RuntimeError):
+        pass
+
+    class Solitary(Unreachable):
+        message = "Cannot connect to any teacher nodes."
+
+    class Lonely(Unreachable):
+        message = "Cannot connect to enough teacher nodes."
+
+    def __init__(self, ursula, enforce_loneliness: bool = True):
+
+        self.log = Logger(self.__class__.__name__)
+        self._ursula = ursula
+        self.enforce_loneliness = enforce_loneliness
+
+        self.__score = 10
+        # 10 == Fully available
+        self.warnings = {
+            9: self.mild_warning,
+            7: self.medium_warning,
+            2: self.severe_warning,
+            0: self.shutdown_everything
+        }
+
+        self._start_time = None
+        self.__task = LoopingCall(self.maintain)
+
+    def mild_warning(self) -> None:
+        self.log.info(f'[UNREACHABLE NOTICE] This node was recently reported as unreachable.')
+
+    def medium_warning(self) -> None:
+        self.log.warn(f'[UNREACHABLE CAUTION] This node is reporting as unreachable.'
+                      f'Please check your network and firewall configuration.')
+
+    def severe_warning(self) -> None:
+        # TODO: Event hooks
+        self.log.warn(f'[UNREACHABLE WARNING] '
+                      f'Please check your network and firewall configuration.'
+                      f'Auto-shutdown will commence soon if the services do not become available.')
+
+    def shutdown_everything(self, reason=None):
+        try:
+            if reason:
+                raise reason(reason.message)
+            raise self.Unreachable(f'{self._ursula} is unreachable.')
+        finally:
+            if reactor.running:
+                reactor.stop()
+
+    def handle_measurement_errors(self, *args, **kwargs) -> None:
+        failure = args[0]
+        cleaned_traceback = failure.getTraceback().replace('{', '').replace('}', '')  # FIXME: Amazing.
+        self.log.warn("Unhandled error during availability check: {}".format(cleaned_traceback))
+        failure.raiseException()
+
+    def status(self) -> bool:
+        """Returns current indication of availability"""
+        return self.score > self.SENSOR_SENSITIVITY
+
+    @property
+    def running(self) -> bool:
+        return self.__task.running
+
+    def start(self, now: bool = False):
+        if not self.running:
+            self._start_time = maya.now()
+            d = self.__task.start(interval=self.FAST_INTERVAL, now=now)
+            d.addErrback(self.handle_measurement_errors)
+
+    def stop(self) -> None:
+        if self.running:
+            self.__task.stop()
+
+    def maintain(self) -> None:
+        self.log.debug(f"Measuring node availability current score is {self.score}.")
+        known_nodes_is_smaller_than_sample_size = len(self._ursula.known_nodes) < self.SAMPLE_SIZE
```

Comment:
Does a sample size of 1 make sense? Is this statement likely, since the default teacher node is our own? I'm sure that I'm missing something here - just need a little clarification

Reply:
This is just a simple optimization mechanism, anticipating some nodes being incompatible.